### PR TITLE
Feature/fs 14

### PR DESCRIPTION
**Pull request by Deizerok**
https://kanbanflow.com/t/LAM3DmRA

### CODE REVIEW

File: app/src/main/java/com/example/fakestore/content/products/presentation/ProductPositionLiveDataWrapper.kt [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+interface ProductPositionLiveDataWrapper {
+
+    interface Update : UiUpdate<Int>
+
+    interface Provide : ProvideLiveData<Int>
+
+    interface Mutable : Update, Provide
+
+    class Base @Inject constructor() : Mutable, LiveDataWrapper.Single<Int>() {
+        init {
```
Comment (bugbenderx) [inline app/src/main/java/com/example/fakestore/content/products/presentation/ProductPositionLiveDataWrapper.kt]: remove init

File: app/src/main/java/com/example/fakestore/content/products/presentation/ProductsViewModel.kt [hunk 2]
```diff
@@ -34,31 +43,38 @@ class ProductsViewModel @Inject constructor(
...
+            val productUI = productItem.map(productItemToProductUiMapper)
+            productUiCommunication.updateUi(productUI)
+        }
+    }
+    
     override fun retry(category: String) {
         init(category)
     }
 
     override fun goToProductsDetails(id: Int, category: String) {
```
Comment (bugbenderx) [inline app/src/main/java/com/example/fakestore/content/products/presentation/ProductsViewModel.kt]: no need category

File: app/src/test/java/com/example/fakestore/content/products/presentation/ProductDetailsViewModelTest.kt [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+                count = 4,
+                favorite = false,
+                addedToCart = false
+            )
+        )
+    }
+
+    @Test
+    fun testBackToProducts() {
+        viewModel.goToProducts(1)
```
Comment (bugbenderx) [inline app/src/test/java/com/example/fakestore/content/products/presentation/ProductDetailsViewModelTest.kt]: add explicit name argument

Comment (bugbenderx) [reply app/src/test/java/com/example/fakestore/content/products/presentation/ProductDetailsViewModelTest.kt]: check value in productPositionLiveDataWrapper

File: app/src/test/java/com/example/fakestore/content/products/presentation/ProductDetailsViewModelTest.kt [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+    override suspend fun changeFavorite(id: Int) {
+        if (id == 1) {
+            val oldCache = cache
+            val newCache = cache.copy(favorite = true)
+            cache = newCache
+        }
+    }
+
+    override suspend fun changeAddedToCart(id: Int): Int {
```
Comment (bugbenderx) [inline app/src/test/java/com/example/fakestore/content/products/presentation/ProductDetailsViewModelTest.kt:194]: override suspend fun changeAddedToCart(id: Int): Int {
        return if (changeAddedToCartCalledCount++ %2 == 0) {
            cache = cache.copy(addedToCart = true)
            1
        }else {
            cache = cache.copy(addedToCart = false)
            0
        }
    }

File: app/src/test/java/com/example/fakestore/content/products/presentation/ProductDetailsViewModelTest.kt [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+        count = 4,
+        favorite = false,
+        addedToCart = false
+    )
+
+    override suspend fun product(id: Int): ProductItem {
+        return cache
+    }
+
+    override suspend fun changeFavorite(id: Int) {
```
Comment (bugbenderx) [inline app/src/test/java/com/example/fakestore/content/products/presentation/ProductDetailsViewModelTest.kt:178]: see override suspend fun changeAddedToCart but without return Int

**Review by bugbenderx — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/fakestore/content/products/presentation/ProductsViewModel.kt (hunk 2 above)
Comment (Deizerok) [reply app/src/main/java/com/example/fakestore/content/products/presentation/ProductsViewModel.kt]: geniy how do i make a retry without a category

**Review by bugbenderx — APPROVED** [review]

**Pull request merged**